core/src/main/java/edu/wpi/grip/core/http/GripServer.java

```diff
@@ -0,0 +1,404 @@
+public class GripServer {
+
+    private static final Logger logger = Logger.getLogger(GripServer.class.getName());
+
+    /**
+     * The root path for all GRIP-related HTTP activity.
+     */
+    public static final String ROOT_PATH = "/GRIP";
+
+    /**
+     * The root path for uploading data to the server.
+     */
+    public static final String UPLOAD_PATH = ROOT_PATH + "/upload";
+
+    /**
+     * The path for uploading images. To upload an image, post an HTTP event to
+     * {@code http://$grip_ip:$grip_port/GRIP/upload/image} (where
+     * {@code $grip_ip} is the IP address of the machine that GRIP is running
+     * on, and {@code $grip_port} is the port on that machine that this server
+     * is running on), with the image bytes as the data.
+     */
+    public static final String IMAGE_UPLOAD_PATH = UPLOAD_PATH + "/image";
+
+    /**
+     * The path for setting which pipeline to run. To set the pipeline, post an
+     * HTTP event to {@code http://$grip_ip:$grip_port/GRIP/upload/pipeline}
+     * with the name of the pipeline to run as the data.
+     */
+    public static final String PIPELINE_UPLOAD_PATH = UPLOAD_PATH + "/pipeline";
+
+    /**
+     * The path for requesting data. Data will be returned as a json-formatted
+     * map of the outputs of all requested data sets.
+     * <p>
+     * For example, performing a {@code GET} request on the path
+     * {@code /GRIP/data?foo;bar} will return a map such as
+     * <br>
+     * <code><pre>
+     * {
+     *   'foo': {
+     *        // data
+     *      },
+     *   'bar':
+     *     {
+     *       // data
+     *     }
+     * }
+     * </pre></code>
+     */
+    public static final String DATA_PATH = ROOT_PATH + "/data";
+
+    /* HTTP constants */
+    private static final String HOSTNAME = "localhost";
+    private static final int BACKLOG = 1;
+
+    private static final String HEADER_ALLOW = "Allow";
+    private static final String HEADER_CONTENT_TYPE = "Content-Type";
+
+    private static final Charset CHARSET = StandardCharsets.UTF_8;
+
+    private static final int STATUS_OK = 200;
+    private static final int STATUS_METHOD_NOT_ALLOWED = 405;
+    private static final int STATUS_INTERNAL_ERROR = 500;
+
+    private static final int NO_RESPONSE_LENGTH = -1;
+
+    private static final String METHOD_GET = "GET";
+    private static final String METHOD_POST = "POST";
+    private static final String METHOD_OPTIONS = "OPTIONS";
+    private static final String ALLOWED_METHODS = String.join(", ", METHOD_GET, METHOD_POST, METHOD_OPTIONS);
+
+    public interface HttpServerFactory {
+        HttpServer create(int port);
+    }
+
+    public static class HttpServerFactoryImpl implements HttpServerFactory {
+        @Override
+        public HttpServer create(int port) {
+            try {
+                return HttpServer.create(new InetSocketAddress(HOSTNAME, port), BACKLOG);
+            } catch (IOException ex) {
+                throw new GripException(ex);
```

Comment:
`IllegalStateException` "Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation."

There doesn't seem to be an applicable `RuntimeException` subclass in the java libraries